src/storage-client/src/source/persist_source.rs

```diff
@@ -528,18 +588,270 @@ impl PersistSourceDataStats<'_> {
+                            // have emitted an entire `summary` worth of data, and can store this
+                            // value for later.
+                            if inflight_bytes >= flow_control_max_bytes
+                                && !PartialOrder::less_than(
+                                    &output_frontier,
+                                    &flow_control_frontier,
+                                )
+                            {
+                                pending_parts.push_back((time, part));
+                                continue 'emitting_parts;
+                            }
+                            (time, part)
+                        }
+                        Some(Either::Left(prog)) => {
+                            output_frontier = prog;
+                            cap_set.downgrade(output_frontier.iter());
+                            continue 'emitting_parts;
+                        }
+                        None => {
+                            if pending_parts.is_empty() {
+                                break 'emitting_parts;
+                            } else {
+                                continue 'emitting_parts;
+                            }
+                        }
+                    }
+                };
+
+                // Store the value with the _frontier_ the `flow_control_input` must reach
+                // to retire it. Note that if this `results_in` is `None`, then we
+                // are at `T::MAX`, and give up on flow_control entirely.
+                //
+                // SUBTLE: If we stop storing these parts, we will likely never check the
+                // `flow_control_input` ever again. This won't pile up data as that input
+                // only has frontier updates. There may be spurious activations from it though.
+                //
+                // Also note that we don't attempt to handle overflowing the `u64` part counter.
+                if let Some(emission_ts) = flow_control.summary.results_in(&time) {
+                    inflight_parts.push((emission_ts, part.byte_size()));
+                }
+
+                // Emit the data at the given time, and update the frontier and capabilities
+                // to just beyond the part.
+                data_output.give(&cap_set.delayed(&time), part).await;
+                let mut just_beyond = time.clone();
+                just_beyond.1 += 1;
+                output_frontier = Antichain::from_elem(just_beyond);
+                cap_set.downgrade(output_frontier.iter())
+            } else {
+                let parts_count = inflight_parts.len();
+                // We've exhausted our budget, listen for updates to the flow_control
+                // input's frontier until we free up new budget. Progress ChangeBatches
+                // are consumed even if you don't interact with the handle, so even if
```

Comment:
Can you explain what this means? Which `ChangeBatches` are consumed if you don't interact with a handle?

Reply:
clarified it!